src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,658 @@
+                    case QUIC_STREAM_EVENT.RECEIVE:
+                        {
+                            status = HandleEventRecv(ref evt);
+                        }
+                        break;
+                    // Send has completed.
+                    // Contains a canceled bool to indicate if the send was canceled.
+                    case QUIC_STREAM_EVENT.SEND_COMPLETE:
+                        {
+                            status = HandleEventSendComplete(ref evt);
+                        }
+                        break;
+                    // Peer has told us to shutdown the reading side of the stream.
+                    case QUIC_STREAM_EVENT.PEER_SEND_SHUTDOWN:
+                        {
+                            status = HandleEventPeerSendShutdown();
+                        }
+                        break;
+                    // Peer has told us to abort the reading side of the stream.
+                    case QUIC_STREAM_EVENT.PEER_SEND_ABORTED:
+                        {
+                            status = HandleEventPeerSendAborted();
+                        }
+                        break;
+                    // Peer has stopped receiving data, don't send anymore.
+                    // Potentially throw when WriteAsync/FlushAsync.
+                    case QUIC_STREAM_EVENT.PEER_RECEIVE_ABORTED:
+                        {
+                            status = HandleEventPeerRecvAbort();
+                        }
+                        break;
+                    // Occurs when shutdown is completed for the send side.
+                    // This only happens for shutdown on sending, not receiving
+                    // Receive shutdown can only be abortive.
+                    case QUIC_STREAM_EVENT.SEND_SHUTDOWN_COMPLETE:
+                        {
+                            status = HandleEventSendShutdownComplete(ref evt);
+                        }
+                        break;
+                    // Shutdown for both sending and receiving is completed.
+                    case QUIC_STREAM_EVENT.SHUTDOWN_COMPLETE:
+                        {
+                            status = HandleEventShutdownComplete();
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            }
+            catch (Exception)
+            {
+                return MsQuicConstants.InternalError;
+            }
+
+            return status;
+        }
+
+        private unsafe uint HandleEventRecv(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            StreamEventDataRecv receieveEvent = evt.Data.Recv;
+            _quicBuffer = new QuicBuffer[receieveEvent.BufferCount];
+            for (int i = 0; i < receieveEvent.BufferCount; i++)
+            {
+                _quicBuffer[i] = receieveEvent.Buffers[i];
+            }
```

Comment:
Can we ever get a second receive event before our prior buffers are done being consumed?

Can we make `_quicBuffer` into a `List` or `Queue` to avoid reallocating every time we have new buffers?